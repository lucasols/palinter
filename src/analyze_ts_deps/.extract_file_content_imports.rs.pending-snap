{"run_id":"1679511834-285920000","line":593,"new":{"module_name":"palinter__analyze_ts_deps__extract_file_content_imports__tests","snapshot_name":"bug_test","metadata":{"source":"src/analyze_ts_deps/extract_file_content_imports.rs","assertion_line":593,"expression":"imports"},"snapshot":"[\n    Import {\n        import_path: \"@src/api/apiCall\",\n        line: 2,\n        values: Named(\n            [\n                \"apiMutation\",\n            ],\n        ),\n    },\n    Import {\n        import_path: \"@src/api/apiStores/apiStores.utils\",\n        line: 3,\n        values: Named(\n            [\n                \"validateApiResponse\",\n            ],\n        ),\n    },\n    Import {\n        import_path: \"@src/api/schemas/resources/conversationResource\",\n        line: 4,\n        values: Named(\n            [\n                \"ChatType\",\n            ],\n        ),\n    },\n    Import {\n        import_path: \"@src/state/userStore\",\n        line: 5,\n        values: Named(\n            [\n                \"getCurrentUserId\",\n            ],\n        ),\n    },\n    Import {\n        import_path: \"@src/stores/chat/chatMessagesList\",\n        line: 6,\n        values: Named(\n            [\n                \"chatMessagesList\",\n            ],\n        ),\n    },\n    Import {\n        import_path: \"@src/stores/chat/chatsList\",\n        line: 7,\n        values: Named(\n            [\n                \"ChatConversation\",\n                \"chatsList\",\n                \"conversationListMetaId\",\n                \"conversationSchema\",\n                \"convertFromApiConversation\",\n            ],\n        ),\n    },\n    Import {\n        import_path: \"@src/utils/betaFeatures\",\n        line: 14,\n        values: Named(\n            [\n                \"isBetaFeature\",\n            ],\n        ),\n    },\n    Import {\n        import_path: \"@src/utils/history\",\n        line: 15,\n        values: Named(\n            [\n                \"navigate\",\n            ],\n        ),\n    },\n    Import {\n        import_path: \"@utils/typeAssertions\",\n        line: 16,\n        values: Named(\n            [\n                \"strictAssertIsNotNullish\",\n            ],\n        ),\n    },\n]"},"old":{"module_name":"palinter__analyze_ts_deps__extract_file_content_imports__tests","metadata":{},"snapshot":"[\n    Import {\n        import_path: \"@src/api/apiCall\",\n        line: 2,\n        values: Named(\n            [\n                \"apiMutation\",\n            ],\n        ),\n    },\n    Import {\n        import_path: \"@src/api/apiStores/apiStores.utils\",\n        line: 3,\n        values: Named(\n            [\n                \"validateApiResponse\",\n            ],\n        ),\n    },\n    Import {\n        import_path: \"@src/api/schemas/resources/conversationResource\",\n        line: 4,\n        values: Named(\n            [\n                \"ChatType\",\n            ],\n        ),\n    },\n    Import {\n        import_path: \"@src/state/userStore\",\n        line: 5,\n        values: Named(\n            [\n                \"getCurrentUserId\",\n            ],\n        ),\n    },\n    Import {\n        import_path: \"@src/stores/chat/chatMessagesList\",\n        line: 6,\n        values: Named(\n            [\n                \"chatMessagesList\",\n            ],\n        ),\n    },\n    Import {\n        import_path: \"@src/utils/betaFeatures\",\n        line: 14,\n        values: Named(\n            [\n                \"isBetaFeature\",\n            ],\n        ),\n    },\n    Import {\n        import_path: \"@src/utils/history\",\n        line: 15,\n        values: Named(\n            [\n                \"navigate\",\n            ],\n        ),\n    },\n    Import {\n        import_path: \"@utils/typeAssertions\",\n        line: 16,\n        values: Named(\n            [\n                \"strictAssertIsNotNullish\",\n            ],\n        ),\n    },\n]"}}
