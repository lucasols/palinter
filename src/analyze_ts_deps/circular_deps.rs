use crate::{internal_config::Config, load_folder_structure};
use colored::Colorize;

use super::{
    get_file_edges, get_resolved_path, load_used_project_files_deps_info_from_cfg,
    modules_graph::get_node_deps, ALIASES, ROOT_DIR,
};

use std::path::Path;

pub fn get_detailed_file_circular_deps_result(
    file_path: &Path,
    root_dir: &Path,
    config: Config,
) -> Result<(), String> {
    *ALIASES.lock().unwrap() = config
        .ts_config
        .as_ref()
        .map(|c| c.aliases.clone())
        .unwrap_or_default();
    *ROOT_DIR.lock().unwrap() = root_dir.to_str().unwrap().to_string();

    let resolved_path = get_resolved_path(file_path)?.unwrap();

    let root_structure = match load_folder_structure(
        root_dir,
        &config,
        &root_dir.to_path_buf(),
        true,
    ) {
        Ok(root_structure) => root_structure,
        Err(err) => {
            println!("❌ Error loading folder structure: {}", err);
            std::process::exit(1);
        }
    };

    if let Err(err) = load_used_project_files_deps_info_from_cfg(
        &config,
        &root_structure,
        root_dir,
    ) {
        println!("❌ Error getting used files deps info: {}", err);
        std::process::exit(1);
    };

    let result = get_node_deps(
        &resolved_path.to_str().unwrap().to_string(),
        &mut get_file_edges,
        None,
        true,
        true,
    )?;

    if let Some(circular_deps) = result.circular_deps {
        let mut cdeps = circular_deps;

        let original_len = cdeps.len();

        cdeps.truncate(8);

        println!("🔁 Circular deps found:");

        for dep in &cdeps {
            let parts = dep.split(" > ").collect::<Vec<&str>>();

            for (i, part) in parts.iter().enumerate() {
                let part_to_use = if part.starts_with('|') {
                    part.bright_yellow().to_string()
                } else {
                    part.to_string()
                };

                if i == 0 {
                    println!("\n{} {}", ">>".dimmed(), part_to_use);
                } else {
                    println!("   {}{}", " ".repeat(i * 2), part_to_use);
                }
            }
        }

        if cdeps.len() < original_len {
            println!(
                "{}",
                format!("\n... and {} more", original_len - cdeps.len())
                    .bright_yellow()
            );
        }
        Ok(())
    } else {
        println!("No circular deps found");
        Ok(())
    }
}
