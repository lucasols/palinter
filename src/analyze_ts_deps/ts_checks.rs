use std::collections::HashMap;

use crate::load_folder_structure::File;

use super::{extract_file_content_imports::ImportType, FileDepsInfo};

pub fn check_ts_not_have_unused_exports(
    file: &File,
    used_files_deps_info: &HashMap<String, FileDepsInfo>,
) -> Result<(), String> {
    let deps_info = used_files_deps_info.get(&file.path);

    if let Some(deps_info) = deps_info {
        let mut unused_exports = deps_info.exports.clone();

        for (other_used_file, other_deps_info) in used_files_deps_info {
            if unused_exports.is_empty() {
                break;
            }

            if other_used_file == &file.path {
                continue;
            }

            if let Some(related_import) = other_deps_info.imports.get(&file.path) {
                match &related_import.values {
                    ImportType::All | ImportType::Dynamic => {
                        unused_exports = vec![];
                    }
                    ImportType::Named(values) => {
                        unused_exports
                            .retain(|export| !values.contains(&export.name));
                    }
                    ImportType::SideEffect => {}
                }
            }
        }

        if !unused_exports.is_empty() {
            return Err(format!(
                "File has unused exports: {}",
                unused_exports
                    .iter()
                    .map(|export| format!(
                        "'{}' in line {}",
                        export.name, export.line
                    ))
                    .collect::<Vec<String>>()
                    .join(", ")
            ));
        } else {
            Ok(())
        }
    } else {
        Err("File is not being used in the project".to_string())
    }
}
