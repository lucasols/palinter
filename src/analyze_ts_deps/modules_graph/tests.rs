use crate::analyze_ts_deps::_setup_test;

use super::*;
use indexmap::IndexMap;
use insta::assert_debug_snapshot;

fn vc(v: Vec<&str>) -> Vec<String> {
    v.iter().map(|s| s.to_string()).collect()
}

const EMPTY: Vec<String> = vec![];

fn get_deps_for_each(
    nodes: Vec<&str>,
    mut get_node_edges: impl Fn(&str) -> Result<Vec<String>, String>,
) -> IndexMap<String, DepsResult> {
    nodes
        .iter()
        .map(|node| {
            (
                node.to_string(),
                get_node_deps(&node.to_string(), &mut get_node_edges, Some(1000))
                    .unwrap(),
            )
        })
        .collect()
}

#[test]
fn simple_dep_calc_1() {
    _setup_test();

    let get_node_edges = |node_name: &str| -> Result<Vec<String>, String> {
        Ok(match node_name {
            "dep1" => vc(["dep2"].to_vec()),
            "dep2" => vc(["dep3"].to_vec()),
            "dep3" => EMPTY,
            _ => EMPTY,
        })
    };

    assert_debug_snapshot!(
        get_deps_for_each(
            ["dep1", "dep2", "dep3"].to_vec(),
            get_node_edges,
        ),
        @r###"
    {
        "dep1": DepsResult {
            deps: {
                "dep2",
                "dep3",
            },
            circular_deps: None,
        },
        "dep2": DepsResult {
            deps: {
                "dep3",
            },
            circular_deps: None,
        },
        "dep3": DepsResult {
            deps: {},
            circular_deps: None,
        },
    }
    "###
    )
}

#[test]
fn simple_dep_calc_2() {
    _setup_test();

    let get_node_edges = |node_name: &str| -> Result<Vec<String>, String> {
        Ok(match node_name {
            "dep1" => vc(["dep2", "dep4"].to_vec()),
            "dep2" => vc(["dep3"].to_vec()),
            "dep3" => EMPTY,
            "dep4" => vc(["dep5"].to_vec()),
            "dep5" => EMPTY,
            _ => EMPTY,
        })
    };

    assert_debug_snapshot!(
        get_deps_for_each(
            ["dep1", "dep2", "dep3", "dep4", "dep5"].to_vec(),
            get_node_edges,
        ),
        @r###"
        {
            "dep1": DepsResult {
                deps: {
                    "dep2",
                    "dep3",
                    "dep4",
                    "dep5",
                },
                circular_deps: None,
            },
            "dep2": DepsResult {
                deps: {
                    "dep3",
                },
                circular_deps: None,
            },
            "dep3": DepsResult {
                deps: {},
                circular_deps: None,
            },
            "dep4": DepsResult {
                deps: {
                    "dep5",
                },
                circular_deps: None,
            },
            "dep5": DepsResult {
                deps: {},
                circular_deps: None,
            },
        }
        "###
    )
}

#[test]
fn circular_dep_calc() {
    _setup_test();

    let get_node_edges = |node_name: &str| -> Result<Vec<String>, String> {
        Ok(match node_name {
            "circular" => vc(["dep1"].to_vec()),
            "dep1" => vc(["dep2"].to_vec()),
            "dep2" => vc(["circular"].to_vec()),
            _ => EMPTY,
        })
    };

    assert_debug_snapshot!(
        get_deps_for_each(
            ["circular", "dep1", "dep2"].to_vec(),
            get_node_edges,
        ),
        @r###"
        {
            "circular": DepsResult {
                deps: {
                    "dep1",
                    "dep2",
                    "circular",
                },
                circular_deps: Some(
                    [
                        "|circular| > dep1 > dep2 > |circular|",
                    ],
                ),
            },
            "dep1": DepsResult {
                deps: {
                    "dep2",
                    "circular",
                    "dep1",
                },
                circular_deps: Some(
                    [
                        "|dep1| > dep2 > circular > |dep1|",
                    ],
                ),
            },
            "dep2": DepsResult {
                deps: {
                    "circular",
                    "dep1",
                    "dep2",
                },
                circular_deps: Some(
                    [
                        "|dep2| > circular > dep1 > |dep2|",
                    ],
                ),
            },
        }
        "###
    );
}

#[test]
fn circular_1() {
    _setup_test();

    let get_node_edges = |node_name: &str| -> Result<Vec<String>, String> {
        Ok(match node_name {
            "circular" => vc(["dep1"].to_vec()),
            "dep1" => vc(["dep2", "dep3"].to_vec()),
            "dep2" => vc(["circular"].to_vec()),
            "dep3" => EMPTY,
            _ => EMPTY,
        })
    };

    assert_debug_snapshot!(
        get_deps_for_each(
            ["circular", "dep1", "dep2", "dep3"].to_vec(),
            get_node_edges,
        ),
        @r###"
        {
            "circular": DepsResult {
                deps: {
                    "dep1",
                    "dep2",
                    "circular",
                    "dep3",
                },
                circular_deps: Some(
                    [
                        "|circular| > dep1 > dep2 > |circular|",
                    ],
                ),
            },
            "dep1": DepsResult {
                deps: {
                    "dep2",
                    "circular",
                    "dep1",
                    "dep3",
                },
                circular_deps: Some(
                    [
                        "|dep1| > dep2 > circular > |dep1|",
                    ],
                ),
            },
            "dep2": DepsResult {
                deps: {
                    "circular",
                    "dep1",
                    "dep2",
                    "dep3",
                },
                circular_deps: Some(
                    [
                        "|dep2| > circular > dep1 > |dep2|",
                    ],
                ),
            },
            "dep3": DepsResult {
                deps: {},
                circular_deps: None,
            },
        }
        "###
    );
}

#[test]
fn circular_2() {
    _setup_test();

    let get_node_edges = |node_name: &str| -> Result<Vec<String>, String> {
        Ok(match node_name {
            "dep1" => vc(["dep2"].to_vec()),
            "dep2" => vc(["dep3"].to_vec()),
            "dep3" => vc(["dep2"].to_vec()),
            _ => EMPTY,
        })
    };

    assert_debug_snapshot!(
        get_deps_for_each(
            ["dep1", "dep2", "dep3"].to_vec(),
            get_node_edges,
        ),
        @r###"
    {
        "dep1": DepsResult {
            deps: {
                "dep2",
                "dep3",
            },
            circular_deps: Some(
                [
                    "dep1 > |dep2| > dep3 > |dep2|",
                ],
            ),
        },
        "dep2": DepsResult {
            deps: {
                "dep3",
                "dep2",
            },
            circular_deps: Some(
                [
                    "|dep2| > dep3 > |dep2|",
                ],
            ),
        },
        "dep3": DepsResult {
            deps: {
                "dep2",
                "dep3",
            },
            circular_deps: Some(
                [
                    "|dep3| > dep2 > |dep3|",
                ],
            ),
        },
    }
    "###
    );
}

#[test]
fn circular_3() {
    _setup_test();

    let get_node_edges = |node_name: &str| -> Result<Vec<String>, String> {
        Ok(match node_name {
            "dep1" => vc(["dep2"].to_vec()),
            "dep2" => vc(["dep3", "dep4"].to_vec()),
            "dep3" => vc(["dep2"].to_vec()),
            "dep4" => vc(["dep2", "dep5"].to_vec()),
            "dep5" => vc(["dep1"].to_vec()),
            _ => EMPTY,
        })
    };

    assert_debug_snapshot!(
        get_deps_for_each(
            ["dep1", "dep2", "dep3", "dep4", "dep5"].to_vec(),
            get_node_edges,
        ),
        @r###"
    {
        "dep1": DepsResult {
            deps: {
                "dep2",
                "dep3",
                "dep4",
                "dep5",
                "dep1",
            },
            circular_deps: Some(
                [
                    "dep1 > |dep2| > dep3 > |dep2|",
                    "dep1 > |dep2| > dep4 > |dep2|",
                    "|dep1| > dep2 > dep4 > dep5 > |dep1|",
                ],
            ),
        },
        "dep2": DepsResult {
            deps: {
                "dep3",
                "dep2",
                "dep4",
                "dep5",
                "dep1",
            },
            circular_deps: Some(
                [
                    "|dep2| > dep3 > |dep2|",
                    "|dep2| > dep4 > |dep2|",
                    "|dep2| > dep4 > dep5 > dep1 > |dep2|",
                ],
            ),
        },
        "dep3": DepsResult {
            deps: {
                "dep2",
                "dep3",
                "dep4",
                "dep5",
                "dep1",
            },
            circular_deps: Some(
                [
                    "|dep3| > dep2 > |dep3|",
                    "dep3 > |dep2| > dep4 > |dep2|",
                    "dep3 > |dep2| > dep4 > dep5 > dep1 > |dep2|",
                ],
            ),
        },
        "dep4": DepsResult {
            deps: {
                "dep2",
                "dep3",
                "dep4",
                "dep5",
                "dep1",
            },
            circular_deps: Some(
                [
                    "dep4 > |dep2| > dep3 > |dep2|",
                    "|dep4| > dep2 > |dep4|",
                ],
            ),
        },
        "dep5": DepsResult {
            deps: {
                "dep1",
                "dep2",
                "dep3",
                "dep4",
                "dep5",
            },
            circular_deps: Some(
                [
                    "dep5 > dep1 > |dep2| > dep3 > |dep2|",
                    "dep5 > dep1 > |dep2| > dep4 > |dep2|",
                    "|dep5| > dep1 > dep2 > dep4 > |dep5|",
                ],
            ),
        },
    }
    "###
    );
}

#[test]
fn circular_4() {
    _setup_test();

    let get_node_edges = |node_name: &str| -> Result<Vec<String>, String> {
        Ok(match node_name {
            "dep1" => vc(["dep2"].to_vec()),
            "dep2" => vc(["dep3", "dep4"].to_vec()),
            "dep3" => vc(["dep2"].to_vec()),
            "dep4" => vc(["dep2", "dep5"].to_vec()),
            "dep5" => vc(["dep3"].to_vec()),
            _ => EMPTY,
        })
    };

    assert_debug_snapshot!(
        get_deps_for_each(
            ["dep1", "dep2", "dep3", "dep4", "dep5"].to_vec(),
            get_node_edges,
        ),
        @r###"
    {
        "dep1": DepsResult {
            deps: {
                "dep2",
                "dep3",
                "dep4",
                "dep5",
            },
            circular_deps: Some(
                [
                    "dep1 > |dep2| > dep3 > |dep2|",
                    "dep1 > |dep2| > dep4 > |dep2|",
                ],
            ),
        },
        "dep2": DepsResult {
            deps: {
                "dep3",
                "dep2",
                "dep4",
                "dep5",
            },
            circular_deps: Some(
                [
                    "|dep2| > dep3 > |dep2|",
                    "|dep2| > dep4 > |dep2|",
                ],
            ),
        },
        "dep3": DepsResult {
            deps: {
                "dep2",
                "dep3",
                "dep4",
                "dep5",
            },
            circular_deps: Some(
                [
                    "|dep3| > dep2 > |dep3|",
                    "dep3 > |dep2| > dep4 > |dep2|",
                ],
            ),
        },
        "dep4": DepsResult {
            deps: {
                "dep2",
                "dep3",
                "dep4",
                "dep5",
            },
            circular_deps: Some(
                [
                    "dep4 > |dep2| > dep3 > |dep2|",
                    "|dep4| > dep2 > |dep4|",
                ],
            ),
        },
        "dep5": DepsResult {
            deps: {
                "dep3",
                "dep2",
                "dep4",
                "dep5",
            },
            circular_deps: Some(
                [
                    "dep5 > |dep3| > dep2 > |dep3|",
                    "dep5 > dep3 > |dep2| > dep4 > |dep2|",
                ],
            ),
        },
    }
    "###
    );
}

#[test]
fn non_cyclic_graph() {
    _setup_test();

    let get_node_edges = |node_name: &str| -> Result<Vec<String>, String> {
        Ok(match node_name {
            "a" => vc(["b", "c"].to_vec()),
            "b" => vc(["d"].to_vec()),
            "c" => vc(["d"].to_vec()),
            "d" => EMPTY,
            _ => EMPTY,
        })
    };

    assert_debug_snapshot!(
        get_deps_for_each(
            ["a", "b", "c", "d"].to_vec(),
            get_node_edges,
        ),
        @r###"
        {
            "a": DepsResult {
                deps: {
                    "b",
                    "d",
                    "c",
                },
                circular_deps: None,
            },
            "b": DepsResult {
                deps: {
                    "d",
                },
                circular_deps: None,
            },
            "c": DepsResult {
                deps: {
                    "d",
                },
                circular_deps: None,
            },
            "d": DepsResult {
                deps: {},
                circular_deps: None,
            },
        }
        "###
    );
}

#[test]
fn dag_5() {
    _setup_test();

    let get_node_edges = |node_name: &str| -> Result<Vec<String>, String> {
        Ok(match node_name {
            "dropdown" => vc(["popover", "typings", "uOCO", "uDVU"].to_vec()),
            "popover" => vc(["portalLayer"].to_vec()),
            "portalLayer" => EMPTY,
            "typings" => EMPTY,
            "uOCO" => EMPTY,
            "uDVU" => vc(["useTimout"].to_vec()),
            "useTimout" => vc(["typings"].to_vec()),
            _ => EMPTY,
        })
    };

    assert_debug_snapshot!(
        get_deps_for_each(
            ["dropdown", "popover", "portalLayer", "typings", "uOCO", "uDVU", "useTimout"].to_vec(),
            get_node_edges
        ),
        @r###"
    {
        "dropdown": DepsResult {
            deps: {
                "popover",
                "portalLayer",
                "typings",
                "uOCO",
                "uDVU",
                "useTimout",
            },
            circular_deps: None,
        },
        "popover": DepsResult {
            deps: {
                "portalLayer",
            },
            circular_deps: None,
        },
        "portalLayer": DepsResult {
            deps: {},
            circular_deps: None,
        },
        "typings": DepsResult {
            deps: {},
            circular_deps: None,
        },
        "uOCO": DepsResult {
            deps: {},
            circular_deps: None,
        },
        "uDVU": DepsResult {
            deps: {
                "useTimout",
                "typings",
            },
            circular_deps: None,
        },
        "useTimout": DepsResult {
            deps: {
                "typings",
            },
            circular_deps: None,
        },
    }
    "###
    );
}

#[test]
fn simple_6() {
    _setup_test();

    let get_node_edges = |node_name: &str| -> Result<Vec<String>, String> {
        Ok(match node_name {
            "dep1" => vc(["dep2", "dep4"].to_vec()),
            "dep2" => vc(["dep3"].to_vec()),
            "dep3" => EMPTY,
            "dep4" => vc(["dep2"].to_vec()),
            _ => EMPTY,
        })
    };

    assert_debug_snapshot!(
        get_deps_for_each(
            ["dep1", "dep2", "dep3", "dep4"].to_vec(),
            get_node_edges
        ),
        @r###"
    {
        "dep1": DepsResult {
            deps: {
                "dep2",
                "dep3",
                "dep4",
            },
            circular_deps: None,
        },
        "dep2": DepsResult {
            deps: {
                "dep3",
            },
            circular_deps: None,
        },
        "dep3": DepsResult {
            deps: {},
            circular_deps: None,
        },
        "dep4": DepsResult {
            deps: {
                "dep2",
                "dep3",
            },
            circular_deps: None,
        },
    }
    "###
    );
}

#[test]
fn simple_7() {
    _setup_test();

    let get_node_edges = |node_name: &str| -> Result<Vec<String>, String> {
        Ok(match node_name {
            "dep1" => vc(["dep2", "dep4"].to_vec()),
            "dep2" => vc(["dep3"].to_vec()),
            "dep3" => EMPTY,
            "dep4" => vc(["dep5"].to_vec()),
            "dep5" => EMPTY,
            _ => EMPTY,
        })
    };

    assert_debug_snapshot!(
        get_deps_for_each(
            ["dep1", "dep2", "dep3", "dep4", "dep5"].to_vec(),
            get_node_edges
        ),
        @r###"
            {
                "dep1": DepsResult {
                    deps: {
                        "dep2",
                        "dep3",
                        "dep4",
                        "dep5",
                    },
                    circular_deps: None,
                },
                "dep2": DepsResult {
                    deps: {
                        "dep3",
                    },
                    circular_deps: None,
                },
                "dep3": DepsResult {
                    deps: {},
                    circular_deps: None,
                },
                "dep4": DepsResult {
                    deps: {
                        "dep5",
                    },
                    circular_deps: None,
                },
                "dep5": DepsResult {
                    deps: {},
                    circular_deps: None,
                },
            }
            "###);
}

#[test]
fn circular_8() {
    _setup_test();

    let get_node_edges = |node_name: &str| -> Result<Vec<String>, String> {
        Ok(match node_name {
            "circular" => vc(["dep1"].to_vec()),
            "dep1" => vc(["dep2", "dep3"].to_vec()),
            "dep2" => vc(["circular"].to_vec()),
            "dep3" => vc(["dep4"].to_vec()),
            "dep4" => EMPTY,
            _ => EMPTY,
        })
    };

    assert_debug_snapshot!(
        get_deps_for_each(
            ["circular", "dep1", "dep2", "dep3", "dep4"].to_vec(),
            get_node_edges
        ),
        @r###"
            {
                "circular": DepsResult {
                    deps: {
                        "dep1",
                        "dep2",
                        "circular",
                        "dep3",
                        "dep4",
                    },
                    circular_deps: Some(
                        [
                            "|circular| > dep1 > dep2 > |circular|",
                        ],
                    ),
                },
                "dep1": DepsResult {
                    deps: {
                        "dep2",
                        "circular",
                        "dep1",
                        "dep3",
                        "dep4",
                    },
                    circular_deps: Some(
                        [
                            "|dep1| > dep2 > circular > |dep1|",
                        ],
                    ),
                },
                "dep2": DepsResult {
                    deps: {
                        "circular",
                        "dep1",
                        "dep2",
                        "dep3",
                        "dep4",
                    },
                    circular_deps: Some(
                        [
                            "|dep2| > circular > dep1 > |dep2|",
                        ],
                    ),
                },
                "dep3": DepsResult {
                    deps: {
                        "dep4",
                    },
                    circular_deps: None,
                },
                "dep4": DepsResult {
                    deps: {},
                    circular_deps: None,
                },
            }
            "###);
}

#[test]
fn circular_9() {
    _setup_test();

    let get_node_edges = |node_name: &str| -> Result<Vec<String>, String> {
        Ok(match node_name {
            "A" => vc(["B", "C"].to_vec()),
            "B" => vc(["D", "E"].to_vec()),
            "C" => vc(["D"].to_vec()),
            "D" => vc(["1"].to_vec()),
            "E" => vc(["F"].to_vec()),
            "F" => vc(["B"].to_vec()),
            "1" => vc(["2"].to_vec()),
            "2" => vc(["3"].to_vec()),
            "3" => vc(["1"].to_vec()),
            _ => EMPTY,
        })
    };

    assert_debug_snapshot!(
        get_deps_for_each(
            ["A", "B", "C", "D", "E", "F", "1", "2", "3"].to_vec(),
            get_node_edges
        ),
        @r###"
    {
        "A": DepsResult {
            deps: {
                "B",
                "D",
                "1",
                "2",
                "3",
                "E",
                "F",
                "C",
            },
            circular_deps: Some(
                [
                    "A > B > D > |1| > 2 > 3 > |1|",
                    "A > |B| > E > F > |B|",
                ],
            ),
        },
        "B": DepsResult {
            deps: {
                "D",
                "1",
                "2",
                "3",
                "E",
                "F",
                "B",
            },
            circular_deps: Some(
                [
                    "B > D > |1| > 2 > 3 > |1|",
                    "|B| > E > F > |B|",
                ],
            ),
        },
        "C": DepsResult {
            deps: {
                "D",
                "1",
                "2",
                "3",
            },
            circular_deps: Some(
                [
                    "C > D > |1| > 2 > 3 > |1|",
                ],
            ),
        },
        "D": DepsResult {
            deps: {
                "1",
                "2",
                "3",
            },
            circular_deps: Some(
                [
                    "D > |1| > 2 > 3 > |1|",
                ],
            ),
        },
        "E": DepsResult {
            deps: {
                "F",
                "B",
                "D",
                "1",
                "2",
                "3",
                "E",
            },
            circular_deps: Some(
                [
                    "E > F > B > D > |1| > 2 > 3 > |1|",
                    "|E| > F > B > |E|",
                ],
            ),
        },
        "F": DepsResult {
            deps: {
                "B",
                "D",
                "1",
                "2",
                "3",
                "E",
                "F",
            },
            circular_deps: Some(
                [
                    "F > B > D > |1| > 2 > 3 > |1|",
                    "|F| > B > E > |F|",
                ],
            ),
        },
        "1": DepsResult {
            deps: {
                "2",
                "3",
                "1",
            },
            circular_deps: Some(
                [
                    "|1| > 2 > 3 > |1|",
                ],
            ),
        },
        "2": DepsResult {
            deps: {
                "3",
                "1",
                "2",
            },
            circular_deps: Some(
                [
                    "|2| > 3 > 1 > |2|",
                ],
            ),
        },
        "3": DepsResult {
            deps: {
                "1",
                "2",
                "3",
            },
            circular_deps: Some(
                [
                    "|3| > 1 > 2 > |3|",
                ],
            ),
        },
    }
    "###);
}

#[test]
fn circular_11() {
    _setup_test();

    let get_node_edges = |node_name: &str| -> Result<Vec<String>, String> {
        Ok(match node_name {
            "A" => vc(["B", "C"].to_vec()),
            "B" => vc(["D", "E"].to_vec()),
            "C" => vc(["1", "3"].to_vec()),
            "D" => EMPTY,
            "E" => vc(["F"].to_vec()),
            "F" => vc(["E"].to_vec()),
            "3" => EMPTY,
            "1" => vc(["2"].to_vec()),
            "2" => vc(["1"].to_vec()),
            _ => EMPTY,
        })
    };

    assert_debug_snapshot!(get_deps_for_each(
        ["A", "B", "C", "D", "E", "F", "1", "2", "3"].to_vec(),
        get_node_edges
    ), @r###"
    {
        "A": DepsResult {
            deps: {
                "B",
                "D",
                "E",
                "F",
                "C",
                "1",
                "2",
                "3",
            },
            circular_deps: Some(
                [
                    "A > B > |E| > F > |E|",
                    "A > C > |1| > 2 > |1|",
                ],
            ),
        },
        "B": DepsResult {
            deps: {
                "D",
                "E",
                "F",
            },
            circular_deps: Some(
                [
                    "B > |E| > F > |E|",
                ],
            ),
        },
        "C": DepsResult {
            deps: {
                "1",
                "2",
                "3",
            },
            circular_deps: Some(
                [
                    "C > |1| > 2 > |1|",
                ],
            ),
        },
        "D": DepsResult {
            deps: {},
            circular_deps: None,
        },
        "E": DepsResult {
            deps: {
                "F",
                "E",
            },
            circular_deps: Some(
                [
                    "|E| > F > |E|",
                ],
            ),
        },
        "F": DepsResult {
            deps: {
                "E",
                "F",
            },
            circular_deps: Some(
                [
                    "|F| > E > |F|",
                ],
            ),
        },
        "1": DepsResult {
            deps: {
                "2",
                "1",
            },
            circular_deps: Some(
                [
                    "|1| > 2 > |1|",
                ],
            ),
        },
        "2": DepsResult {
            deps: {
                "1",
                "2",
            },
            circular_deps: Some(
                [
                    "|2| > 1 > |2|",
                ],
            ),
        },
        "3": DepsResult {
            deps: {},
            circular_deps: None,
        },
    }
    "###);
}
