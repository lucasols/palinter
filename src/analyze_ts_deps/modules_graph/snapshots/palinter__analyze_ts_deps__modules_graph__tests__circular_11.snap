---
source: src/analyze_ts_deps/modules_graph/tests.rs
expression: "get_deps_for_each([\"A\", \"B\", \"C\", \"D\", \"E\", \"F\", \"1\", \"2\", \"3\"].to_vec(),\n    get_node_edges)"
---
{
    "A": DepsResult {
        deps: {
            "A",
            "B",
            "D",
            "F",
            "E",
            "C",
            "1",
            "3",
            "2",
        },
        circular_deps: Some(
            [
                "A > B > |D| > F > |D|",
                "A > E > C > |1| > 3 > |1|",
            ],
        ),
    },
    "B": DepsResult {
        deps: {
            "B",
            "D",
            "F",
            "E",
        },
        circular_deps: Some(
            [
                "B > |D| > F > |D|",
            ],
        ),
    },
    "C": DepsResult {
        deps: {
            "C",
            "1",
            "3",
            "2",
        },
        circular_deps: Some(
            [
                "C > |1| > 3 > |1|",
            ],
        ),
    },
    "D": DepsResult {
        deps: {
            "D",
            "F",
        },
        circular_deps: Some(
            [
                "|D| > F > |D|",
            ],
        ),
    },
    "E": DepsResult {
        deps: {},
        circular_deps: None,
    },
    "F": DepsResult {
        deps: {
            "F",
            "D",
        },
        circular_deps: Some(
            [
                "|F| > D > |F|",
            ],
        ),
    },
    "1": DepsResult {
        deps: {
            "1",
            "3",
        },
        circular_deps: Some(
            [
                "|1| > 3 > |1|",
            ],
        ),
    },
    "2": DepsResult {
        deps: {},
        circular_deps: None,
    },
    "3": DepsResult {
        deps: {
            "3",
            "1",
        },
        circular_deps: Some(
            [
                "|3| > 1 > |3|",
            ],
        ),
    },
}
